File: backend-fastapi/app/services/agent_registry.py
"""
Agent Registry for the NeuroNest AI system.
This module provides a registry for agent types and instances.
"""

import logging
from typing import Dict, List, Any, Optional, Callable, Type

from app.services.base_agent import BaseAgent, EventStream
from app.services.orchestrator_agent import OrchestratorAgent
from app.services.thinking_agent import ThinkingAgent
from app.services.specialized_agents import CodeGenerationAgent, DebuggingAgent, ProjectPlanningAgent

logger = logging.getLogger(__name__)

# Type for agent factory functions
AgentFactory = Callable[[EventStream, Any], BaseAgent]

class AgentRegistry:
    """Registry for agent types and instances"""
    
    _instance = None
    
    def __new__(cls):
        """Ensure singleton pattern"""
        if cls._instance is None:
            cls._instance = super(AgentRegistry, cls).__new__(cls)
            cls._instance.agent_types = {}
            cls._instance.agents = {}
            cls._instance._register_default_agents()
        return cls._instance
    
    @classmethod
    def get_instance(cls) -> 'AgentRegistry':
        """
        Get the singleton instance of the registry
        
        Returns:
            The registry instance
        """
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance
    
    def _register_default_agents(self) -> None:
        """Register default agent types"""
        self.register_agent_type("orchestrator", lambda es, llm: OrchestratorAgent(es, llm))
        self.register_agent_type("thinking", lambda es, llm: ThinkingAgent(es, llm))
        self.register_agent_type("code_generation", lambda es, llm: CodeGenerationAgent(es, llm))
        self.register_agent_type("debugging", lambda es, llm: DebuggingAgent(es, llm))
        self.register_agent_type("project_planning", lambda es, llm: ProjectPlanningAgent(es, llm))
    
    def register_agent_type(self, agent_type: str, factory: AgentFactory) -> None:
        """
        Register an agent type
        
        Args:
            agent_type: Type of the agent
            factory: Factory function to create the agent
        """
        self.agent_types[agent_type] = factory
        logger.info(f"Registered agent type: {agent_type}")
    
    def get_agent_type(self, agent_type: str) -> Optional[AgentFactory]:
        """
        Get an agent type factory
        
        Args:
            agent_type: Type of the agent
            
        Returns:
            The agent factory function
        """
        return self.agent_types.get(agent_type)
    
    def get_agent_types(self) -> List[str]:
        """
        Get all registered agent types
        
        Returns:
            Array of agent type names
        """
        return list(self.agent_types.keys())
    
    async def create_agent(
        self,
        agent_type: str,
        config: Dict[str, Any],
        event_stream: EventStream,
        llm_provider: Any
    ) -> BaseAgent:
        """
        Create an agent instance
        
        Args:
            agent_type: Type of the agent
            config: Configuration for the agent
            event_stream: Event stream for the agent
            llm_provider: LLM provider for the agent
            
        Returns:
            The created agent instance
        """
        factory = self.agent_types.get(agent_type)
        
        if not factory:
            raise ValueError(f"Agent type '{agent_type}' not registered")
        
        agent = factory(event_stream, llm_provider)
        await agent.initialize(config)
        
        self.agents[agent.id] = agent
        logger.info(f"Created agent instance: {agent.id} ({agent_type})")
        
        return agent
    
    def get_agent(self, agent_id: str) -> Optional[BaseAgent]:
        """
        Get an agent instance by ID
        
        Args:
            agent_id: ID of the agent
            
        Returns:
            The agent instance
        """
        return self.agents.get(agent_id)
    
    def get_agents(self) -> List[BaseAgent]:
        """
        Get all agent instances
        
        Returns:
            Array of agent instances
        """
        return list(self.agents.values())
    
    def remove_agent(self, agent_id: str) -> None:
        """
        Remove an agent instance
        
        Args:
            agent_id: ID of the agent
        """
        if agent_id in self.agents:
            logger.info(f"Removed agent instance: {agent_id}")
            del self.agents[agent_id]
    
    def clear_agents(self) -> None:
        """Clear all agent instances"""
        self.agents = {}
        logger.info("Cleared all agent instances")

File: backend/src/agents/AgentRegistry.ts
import { EventStream } from './EventStream';
import { LLMProvider } from './LLMProvider';
import { AgentInterface, AgentConfig } from './AgentInterface';

/**
 * Type for agent factory functions
 */
export type AgentFactory = (eventStream: EventStream, llmProvider: LLMProvider) => AgentInterface;

/**
 * Registry for agent types and instances
 */
export class AgentRegistry {
  private static instance: AgentRegistry;
  private agentTypes: Map<string, AgentFactory> = new Map();
  private agents: Map<string, AgentInterface> = new Map();
  
  /**
   * Get the singleton instance of the registry
   * @returns The registry instance
   */
  public static getInstance(): AgentRegistry {
    if (!AgentRegistry.instance) {
      AgentRegistry.instance = new AgentRegistry();
    }
    return AgentRegistry.instance;
  }
  
  /**
   * Private constructor to enforce singleton pattern
   */
  private constructor() {}
  
  /**
   * Register an agent type
   * @param type Type of the agent
   * @param factory Factory function to create the agent
   */
  registerAgentType(type: string, factory: AgentFactory): void {
    this.agentTypes.set(type, factory);
  }
  
  /**
   * Get an agent type factory
   * @param type Type of the agent
   * @returns The agent factory function
   */
  getAgentType(type: string): AgentFactory | undefined {
    return this.agentTypes.get(type);
  }
  
  /**
   * Get all registered agent types
   * @returns Array of agent type names
   */
  getAgentTypes(): string[] {
    return Array.from(this.agentTypes.keys());
  }
  
  /**
   * Create an agent instance
   * @param type Type of the agent
   * @param config Configuration for the agent
   * @param eventStream Event stream for the agent
   * @param llmProvider LLM provider for the agent
   * @returns The created agent instance
   */
  async createAgent(
    type: string,
    config: AgentConfig,
    eventStream: EventStream,
    llmProvider: LLMProvider
  ): Promise<AgentInterface> {
    const factory = this.agentTypes.get(type);
    
    if (!factory) {
      throw new Error(`Agent type '${type}' not registered`);
    }
    
    const agent = factory(eventStream, llmProvider);
    await agent.initialize(config);
    
    this.agents.set(agent.id, agent);
    
    return agent;
  }
  
  /**
   * Get an agent instance by ID
   * @param id ID of the agent
   * @returns The agent instance
   */
  getAgent(id: string): AgentInterface | undefined {
    return this.agents.get(id);
  }
  
  /**
   * Get all agent instances
   * @returns Array of agent instances
   */
  getAgents(): AgentInterface[] {
    return Array.from(this.agents.values());
  }
  
  /**
   * Remove an agent instance
   * @param id ID of the agent
   */
  removeAgent(id: string): void {
    this.agents.delete(id);
  }
  
  /**
   * Clear all agent instances
   */
  clearAgents(): void {
    this.agents.clear();
  }
}